### PR TITLE
Data collator with padding

### PR DESCRIPTION
**Pull request by sgugger**
Add a data collator to dynamically pad samples during batching. This is necessary for the training set since padding can't be applied beforehand if we use shuffling (unless with pad to a fixed `max_length`).
This should make it more straightforward to plug nlp into the Trainer.

### CODE REVIEW

**Review by LysandreJik — COMMENTED** [review]
This is a welcome addition. Can this collator be used alongside other collators? It seems to be an option that we would like on all of them, given its utility.

**Comment by sgugger** [comment]
It's hard to see how to combine it with another data collator since a data collator's function is to create batch, and you can't create batches if your tensors are not padded to the same size.

File: src/transformers/data/data_collator.py [hunk 1]
```diff
@@ -66,6 +67,55 @@ def default_data_collator(features: List[InputDataClass]) -> Dict[str, torch.Ten
...
+            This is especially useful to enable the use of Tensor Cores on NVIDIA hardware with compute capability
+            >= 7.5 (Volta).
+    """
+
+    tokenizer: Union[PreTrainedTokenizer, PreTrainedTokenizerFast]
+    padding: Union[bool, str, PaddingStrategy] = True
+    max_length: Optional[int] = None
+    pad_to_multiple_of: Optional[int] = None
+
+    def __call__(self, features):
```
Comment (patrickvonplaten) [inline src/transformers/data/data_collator.py:102]: Would be very useful to add type hinting here

Comment (patrickvonplaten) [reply src/transformers/data/data_collator.py:102]: `def __call__(self, features: Dict[str, List[int]]) -> Dict[str, torch.Tensor]` maybe?

File: src/transformers/data/data_collator.py [hunk 2]
```diff
@@ -66,6 +67,55 @@ def default_data_collator(features: List[InputDataClass]) -> Dict[str, torch.Ten
...
+        batch = self.tokenizer.pad(
```
Comment (patrickvonplaten) [inline src/transformers/data/data_collator.py:103]: As it is coded now `features` cannot be of a `dict` of `torch.Tensor` `self.tokenizer.pad(...)` would fail, right?
This is a bit different from some other DataCollators that expect `torch.Tensors` as an input. 

Do we want to go more towards an approach where we preprocess everything in "Python" using the `nlp` library and then use the data collators to transformers a dataset of Python lists to a dataset of torch Tensors?

Comment (sgugger) [reply src/transformers/data/data_collator.py:103]: I disagree that data collators expect `torch.Tensor` as inputs: `default_data_tokenizer` accept both. IMO it's the `DataCollatorForLanguageModeling` and variants that are not great in the sense they don't accept both tensors and list of ints as inputs. On that note, this data collator should also work with tensors, I agree. IMO it's the pad method that should be fixed to accept tensors ultimately.

I think we want to support everything in terms of preprocessing, but `nlp` should be a premium citizen.

Comment (patrickvonplaten) [reply src/transformers/data/data_collator.py:107]: Ok, no I agree with you - I also think we should refactor the other data collators to also accept Python list inputs. 

IMO, a nice workflow is:
1) Download a dataset with `nlp`, this will be in the format `Dict[str, List[int or str]]`. 
2) Then we can use the dataset.map() functionality to process the data to `input_ids`, `attention_mask`, etc....Since `nlp.map(...)` does not allow to return `torch.Tensors`, here we have to stay anyways in Python List format.
3) This dataset can then ideally already be used with the `default_data_tokenizer` because all tokenization methods were applied during the .map() operations. 

Here would be an example of this workflow: https://huggingface.co/patrickvonplaten/bert2bert-cnn_dailymail-fp16#training-script.  (Note that the tensors would not even have to be put into torch tensor format). 
What do you think @sgugger ?

Comment (sgugger) [reply src/transformers/data/data_collator.py:107]: I agree, except step 3 should be `default_data_collator` only when you are padding to a fixed max length. With the shuffle of the training set, you need dynamic padding as provided by the data collator introduced in this PR :-)

**Comment by sgugger** [comment]
Things I tried to fix here were actually addressed by @thomwolf in #6423, so waiting for this PR to be merged before merging this one.

**Comment by sgugger** [comment]
Rebase made the PR unreadable. Opening a new clean one.
